### PR TITLE
[Query] Support Description about Graph Space

### PR DESCRIPTION
**Pull request by darionyaphet**
Reopen  [Support Description about Graph Space](https://github.com/vesoft-inc/nebula/pull/365)

```
nebula> describe space nba
=================================================
| ID | Name | Partition number | Replica Factor |
=================================================
|  1 |  nba |                1 |              1 |
-------------------------------------------------
```

### CODE REVIEW

File: src/graph/DescribeSpaceExecutor.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
+/* Copyright (c) 2018 - present, VE Software Inc. All rights reserved
```
Comment (dutor) [inline src/graph/DescribeSpaceExecutor.cpp]: Please update the copyright header.

Do the same to all of your pending PRs, if any new file were added.

**Comment by darionyaphet** [comment]
jenkins go

File: src/graph/DescribeSpaceExecutor.cpp (hunk 1 above)
Comment (darionyaphet) [reply src/graph/DescribeSpaceExecutor.cpp]: fixed

File: src/graph/DescribeSpaceExecutor.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+Status DescribeSpaceExecutor::prepare() {
+    return Status::OK();;
+}
+
+void DescribeSpaceExecutor::execute() {
+    auto *mc = ectx()->getMetaClient();
+    auto *name = sentence_->spaceName();
+    auto spaceId = ectx()->schemaManager()->toGraphSpaceID(*name);
+    auto spaceItem = mc->getSpace(spaceId).get();
```
Comment (laura-ding) [inline src/graph/DescribeSpaceExecutor.cpp]: please use async

File: src/interface/meta.thrift [hunk 3]
```diff
@@ -58,6 +58,17 @@ struct Pair {
     2: string value,
 }
 
+struct SpaceProperties {
```
Comment (dangleptr) [inline src/interface/meta.thrift]: Well done  👍

File: src/interface/meta.thrift [hunk 4]
```diff
@@ -87,13 +98,12 @@ struct ExecResp {
 
 // Graph space related operations.
 struct CreateSpaceReq {
-    1: string space_name,
-    2: i32 parts_num,
-    3: i32 replica_factor,
+    1: string  space_name,
+    2: i32     partition_num,
```
Comment (dangleptr) [inline src/interface/meta.thrift]: Reuse SpaceProperties

File: src/meta/MetaServiceUtils.cpp [hunk 5]
```diff
@@ -26,17 +26,18 @@ std::string MetaServiceUtils::spaceKey(GraphSpaceID spaceId) {
...
     std::string val;
-    val.reserve(256);
-    val.append(reinterpret_cast<const char*>(&partsNum), sizeof(partsNum));
-    val.append(reinterpret_cast<const char*>(&replicaFactor), sizeof(replicaFactor));
-    val.append(name);
+    apache::thrift::CompactSerializer::serialize(properties, &val);
     return val;
 }
 
+cpp2::SpaceProperties  MetaServiceUtils::parseSpace(folly::StringPiece rawData) {
```
Comment (dangleptr) [inline src/meta/MetaServiceUtils.cpp]: Well done  👍

File: src/meta/MetaServiceUtils.cpp [hunk 6]
```diff
@@ -45,8 +46,16 @@ GraphSpaceID MetaServiceUtils::spaceId(folly::StringPiece rawKey) {
     return *reinterpret_cast<const GraphSpaceID*>(rawKey.data() + kSpacesTable.size());
 }
 
-folly::StringPiece MetaServiceUtils::spaceName(folly::StringPiece rawVal) {
-    return rawVal.subpiece(sizeof(int32_t)*2);
+std::string MetaServiceUtils::spaceName(folly::StringPiece rawVal) {
```
Comment (dangleptr) [inline src/meta/MetaServiceUtils.cpp]: Do we still need the three methods ?

File: src/graph/DescribeSpaceExecutor.cpp (hunk 2 above)
Comment (darionyaphet) [reply src/graph/DescribeSpaceExecutor.cpp]: fixed

File: src/interface/meta.thrift (hunk 4 above)
Comment (darionyaphet) [reply src/interface/meta.thrift]: fixed

File: src/meta/MetaServiceUtils.cpp (hunk 6 above)
Comment (darionyaphet) [reply src/meta/MetaServiceUtils.cpp]: the other method used in test case

**Comment by darionyaphet** [comment]
jenkins go

File: src/meta/client/MetaClient.cpp [hunk 7]
```diff
@@ -357,10 +357,12 @@ void MetaClient::diff(const std::unordered_map<GraphSpaceID,
...
 MetaClient::createSpace(std::string name, int32_t partsNum, int32_t replicaFactor) {
+    cpp2::SpaceProperties properties;
+    properties.set_space_name(std::move(name));
+    properties.set_partition_num(partsNum);
+    properties.set_replica_factor(replicaFactor);
     cpp2::CreateSpaceReq req;
-    req.set_space_name(std::move(name));
-    req.set_parts_num(partsNum);
-    req.set_replica_factor(replicaFactor);
+    req.set_properties(properties);
```
Comment (dangleptr) [inline src/meta/client/MetaClient.cpp]: move

File: src/meta/MetaServiceUtils.cpp (hunk 6 above)
Comment (dangleptr) [reply src/meta/MetaServiceUtils.cpp]: If so, we'd better remove them in utils

File: src/meta/processors/CreateSpaceProcessor.cpp [hunk 8]
```diff
@@ -11,9 +11,11 @@ namespace meta {
 
 void CreateSpaceProcessor::process(const cpp2::CreateSpaceReq& req) {
     folly::SharedMutex::WriteHolder wHolder(LockUtils::spaceLock());
-    auto spaceRet = getSpaceId(req.get_space_name());
+    auto properties = req.get_properties();
```
Comment (dangleptr) [inline src/meta/processors/CreateSpaceProcessor.cpp]: auto&

File: src/meta/processors/GetSpaceProcessor.cpp [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    auto spaceId = req.get_space_id();
+    std::string spaceKey = MetaServiceUtils::spaceKey(spaceId);
+    auto ret = doGet(std::move(spaceKey));
+    if (!ret.ok()) {
+        LOG(ERROR) << "Get Space SpaceID: " << req.get_space_id() << " not found";
+        resp_.set_code(cpp2::ErrorCode::E_NOT_FOUND);
+        onFinished();
+        return;
+    }
+    auto  properties = MetaServiceUtils::parseSpace(ret.value());
```
Comment (dangleptr) [inline src/meta/processors/GetSpaceProcessor.cpp]: extra space

File: src/meta/test/ProcessorTest.cpp [hunk 10]
```diff
@@ -121,10 +121,12 @@ TEST(ProcessorTest, CreateSpaceTest) {
...
     {
+        cpp2::SpaceProperties properties;
+        properties.set_space_name("default_space");
+        properties.set_partition_num(9);
+        properties.set_replica_factor(3);
         cpp2::CreateSpaceReq req;
-        req.set_space_name("default_space");
-        req.set_parts_num(9);
-        req.set_replica_factor(3);
+        req.set_properties(properties);
```
Comment (dangleptr) [inline src/meta/test/ProcessorTest.cpp]: move

File: src/meta/test/ProcessorTest.cpp [hunk 11]
```diff
@@ -187,10 +189,13 @@ TEST(ProcessorTest, CreateTagTest) {
...
     {
+        cpp2::SpaceProperties properties;
+        properties.set_space_name("default_space");
+        properties.set_partition_num(9);
+        properties.set_replica_factor(3);
         cpp2::CreateSpaceReq req;
-        req.set_space_name("default_space");
-        req.set_parts_num(9);
-        req.set_replica_factor(1);
+        req.set_properties(properties);
```
Comment (dangleptr) [inline src/meta/test/ProcessorTest.cpp]: ditto

File: src/meta/test/ProcessorTest.cpp [hunk 12]
```diff
@@ -235,10 +240,13 @@ TEST(ProcessorTest, CreateEdgeTest) {
...
     {
+        cpp2::SpaceProperties properties;
+        properties.set_space_name("default_space");
+        properties.set_partition_num(9);
+        properties.set_replica_factor(3);
         cpp2::CreateSpaceReq req;
-        req.set_space_name("first_space");
-        req.set_parts_num(9);
-        req.set_replica_factor(1);
+        req.set_properties(properties);
```
Comment (dangleptr) [inline src/meta/test/ProcessorTest.cpp]: ditto

File: src/meta/test/ProcessorTest.cpp [hunk 13]
```diff
@@ -324,10 +335,12 @@ TEST(ProcessorTest, KVOperationTest) {
...
     {
+        cpp2::SpaceProperties properties;
+        properties.set_space_name("default_space");
+        properties.set_partition_num(9);
+        properties.set_replica_factor(3);
         cpp2::CreateSpaceReq req;
-        req.set_space_name("default_space");
-        req.set_parts_num(9);
-        req.set_replica_factor(3);
+        req.set_properties(properties);
```
Comment (dangleptr) [inline src/meta/test/ProcessorTest.cpp]: ditto

File: src/meta/test/ProcessorTest.cpp [hunk 14]
```diff
@@ -1001,10 +1014,12 @@ TEST(ProcessorTest, SameNameTagsTest) {
...
     {
+        cpp2::SpaceProperties properties;
+        properties.set_space_name("default_space");
+        properties.set_partition_num(9);
+        properties.set_replica_factor(3);
         cpp2::CreateSpaceReq req;
-        req.set_space_name("first_space");
-        req.set_parts_num(9);
-        req.set_replica_factor(1);
+        req.set_properties(properties);
```
Comment (dangleptr) [inline src/meta/test/ProcessorTest.cpp]: ditto

File: src/meta/test/ProcessorTest.cpp [hunk 15]
```diff
@@ -1013,10 +1028,12 @@ TEST(ProcessorTest, SameNameTagsTest) {
...
     {
+        cpp2::SpaceProperties properties;
+        properties.set_space_name("second_space");
+        properties.set_partition_num(9);
+        properties.set_replica_factor(1);
         cpp2::CreateSpaceReq req;
-        req.set_space_name("second_space");
-        req.set_parts_num(9);
-        req.set_replica_factor(1);
+        req.set_properties(properties);
```
Comment (dangleptr) [inline src/meta/test/ProcessorTest.cpp]: ditto

File: src/parser/AdminSentences.cpp [hunk 16]
```diff
@@ -49,25 +49,13 @@ std::string HostList::toString() const {
...
 
-
 std::string RemoveHostsSentence::toString() const {
-    std::string buf;
-    buf.reserve(256);
-    buf += "REMOVE HOSTS (";
-    buf += hosts_->toString();
-    buf += ") ";
-    return buf;
+    return folly::stringPrintf("REMOVE HOSTS (%s) ", hosts_->toString().c_str());
```
Comment (dangleptr) [inline src/parser/AdminSentences.cpp]: hosts_->start()

**Pull request merged**